state/state.go

```diff
@@ -119,21 +143,67 @@ func (s *State) LoadABCIResponses() *ABCIResponses {
 	return abciResponses
 }
 
+// SaveValidators persists the validator set for the next block to disk.
+// It should be called after the validator set is updated with the results of EndBlock.
+// If the validator set did not change after processing the latest block,
+// only the last height for which the validators changed is persisted.
+func (s *State) SaveValidators() {
+	lastHeight := s.lastHeightValidatorsChanged
+	nextHeight := s.LastBlockHeight + 1
+	v := &Validators{
+		LastHeightChanged: lastHeight,
+	}
+	if lastHeight == nextHeight {
+		v.ValidatorSet = s.Validators
+	}
+	s.db.SetSync(calcValidatorsKey(nextHeight), v.Bytes())
+}
+
+// LoadValidators loads the ValidatorSet for a given height.
+func (s *State) LoadValidators(height int) (*types.ValidatorSet, error) {
+	v := s.loadValidators(height)
+	if v == nil {
+		return nil, ErrNoValSetForHeight{height}
+	}
+
+	if v.ValidatorSet == nil {
+		v = s.loadValidators(v.LastHeightChanged)
+		if v == nil {
+			return nil, ErrNoValSetForHeight{height}
```

Comment:
maybe this should be a panic or a different type of error

Reply:
if this should not ever happen, then yes, we should panic.

Reply:
+1 for panic if it should never happen.